front/lib/swr/data_source_documents.ts

```diff
@@ -0,0 +1,58 @@
+import type { DataSourceViewType, LightWorkspaceType } from "@dust-tt/types";
+import type {
+  PatchDataSourceWithNameDocumentRequestBody,
+  PostDataSourceWithNameDocumentRequestBody,
+} from "@dust-tt/types";
+import type { SWRMutationConfiguration } from "swr/mutation";
+import useSWRMutation from "swr/mutation";
+
+import { fetcherWithBody } from "@app/lib/swr/swr";
+import type { PostDocumentResponseBody } from "@app/pages/api/w/[wId]/spaces/[spaceId]/data_sources/[dsId]/documents";
+import type { PatchDocumentResponseBody } from "@app/pages/api/w/[wId]/spaces/[spaceId]/data_sources/[dsId]/documents/[documentId]";
+
+async function sendPatchRequest(
+  url: string,
+  {
+    arg,
+  }: {
+    arg: {
+      documentBody: PatchDataSourceWithNameDocumentRequestBody;
+    };
+  }
+) {
+  const res = await fetcherWithBody([url, arg.documentBody, "PATCH"]);
+  return res;
+}
+
+export function useUpdateDataSourceDocumentMutation(
```

Comment:
Any reason to not follow the existing pattern with `doCreate` and `doUpdate`. Usually those hooks will return a method that can be called which abstracts all the mutations.

Reply:
I prefer swr mutations because you get advanced swr features for free, such as optimistic updates, rollback and onSuccess/Error hooks, and fetch management (isMutating). It also makes more sense to me to use swr built in mutation system if we already use it for fetches.

Reply:
It also handle race conditions natively https://swr.vercel.app/docs/mutation#avoid-race-conditions

Reply:
FYI, we have a pattern to return the mutation directly from the fetching hook but disabled from fetching.
For example, here it could use
```js
const {mutateDocument} = useDataSourceViewDocument(.., disabled: true);
```

(weird that you are using datasource and not datasourceview btw).

Reply:
> I prefer swr mutations because you get advanced swr features for free, such as optimistic updates, rollback and onSuccess/Error hooks, and fetch management (isMutating).

Not sure to understand how the `doCreate` and `doUpdate` pattern breaks this. It's just a wrapper to keep the SWR related code in the `swr` folder. You still get all the benefits you highlighted above. Happy to go over this live :).